reactor-test/src/main/java/reactor/test/publisher/ValidatingPublisher.java

```diff
@@ -0,0 +1,437 @@
+package reactor.test.publisher;
+
+import java.util.EnumSet;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+import java.util.stream.Stream;
+
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import reactor.core.publisher.Operators;
+
+/**
+ * A {@link Publisher} that you can directly manipulate, triggering
+ * {@link #next(Object, Object[]) onNext}, {@link #complete() onComplete} and
+ * {@link #error(Throwable) onError} events, for testing purposes.
+ * You can assert the state of the publisher using its {@code expectXXX} methods,
+ * usually inside a {@link reactor.test.StepVerifier}'s
+ * {@link reactor.test.StepVerifier.Step#then(Runnable) then} callback.
+ * <p>
+ * The ValidatingPublisher can also be made more lenient towards the RS spec
+ * and allow "bad behavior" to be performed, as enumerated in {@link Misbehavior}.
+ *
+ * @author Simon Baslé
+ */
+public class ValidatingPublisher<T> implements Publisher<T> {
+
+	/**
+	 * Create a {@link Misbehavior misbehaving} {@link ValidatingPublisher}.
+	 *
+	 * @param first the first allowed {@link Misbehavior}
+	 * @param rest additional optional misbehaviors
+	 * @param <T> the type of the publisher
+	 * @return the new misbehaving {@link ValidatingPublisher}
+	 */
+	public static <T> ValidatingPublisher<T> createMisbehaving(Misbehavior first, Misbehavior... rest) {
+		return new ValidatingPublisher<>(first, rest);
+	}
+
+	/**
+	 * Create a standard {@link ValidatingPublisher}.
+	 *
+	 * @param <T> the type of the publisher
+	 * @return the new {@link ValidatingPublisher}
+	 */
+	public static <T> ValidatingPublisher<T> create() {
+		return new ValidatingPublisher<T>();
+	}
+
+	/**
+	 * Possible misbehavior for a {@link ValidatingPublisher}.
+	 */
+	public enum Misbehavior {
```

Comment:
same here, should the name be changed? also, is the enum enough? a builder could be used as an alternative. This boils down to "do we expect users to often allow a lot of misbehaviors?" in which case enums would maybe be less convenient to use.